File: engine/gd/plans.py
"""Server-side mirror of the frontend's `lib/plans.ts`.

The credit ceiling for every AI-spending operation is enforced HERE,
not (only) on the frontend. The FE was treated as the single line of
defence for a while; any user who could send a bearer token directly
to FastAPI could bypass it. This module + `enforce_credits()` close
that gap.

Keep this file in lockstep with `frontend/lib/plans.ts`. The rules
(plan list, credit allowance per tier, credit costs per action,
effective-plan resolution, hamish founder override) must match exactly
so the FE display + the server gate never disagree.
"""
from __future__ import annotations

import math
import time
from datetime import datetime, timezone
from typing import Iterable, Literal, TypedDict


# ─── Plan identity ────────────────────────────────────────────────────────────

PlanId = Literal[
    "free", "beta", "pro_50", "pro_150", "pro_500", "mega", "enterprise",
]

PRO_TIER_IDS = ("pro_50", "pro_150", "pro_500")
# Stripe subscription statuses that count as paid. Anything else
# (past_due, unpaid, canceled, incomplete, incomplete_expired, paused)
# drops the user back to free until the payment recovers.
PAID_STATUSES = frozenset({"active", "trialing"})


def plan_for(plan_id: str | None) -> PlanId:
    """Resolve a stored plan-id string into the canonical PlanId.
    Tolerates the legacy "pro" string by mapping to the entry tier."""
    p = (plan_id or "").strip().lower()
    if p == "mega":
        return "mega"
    if p == "enterprise":
        return "enterprise"
    if p in PRO_TIER_IDS:
        return p  # type: ignore[return-value]
    if p == "pro":
        return "pro_50"  # legacy pre-split → entry tier
    if p == "beta":
        return "beta"
    return "free"


def is_pro_plan(plan: PlanId) -> bool:
    return plan in PRO_TIER_IDS


# ─── Credit costs ─────────────────────────────────────────────────────────────

# How many of each "thing" one credit covers per calendar cycle. Mirror
# of the FE's CREDIT_COSTS exactly.
LABELLED_IMAGES_PER_CREDIT = 100
UPLOADED_IMAGES_PER_CREDIT = 800
STORED_IMAGES_PER_CREDIT_PER_MONTH = 400


def credits_used(
    *,
    labelled_this_month: int,
    uploaded_this_month: int,
    stored_now: int,
    training_credits_this_period: float = 0.0,
) -> float:
    """Convert raw counters into credit-equivalents. Storage rounds
    UP because a partial-month occupation still burns a credit-month;
    the user is occupying that slot.

    `training_credits_this_period` is already in credit units (the ML
    job billing charges 1 credit per completed 15-min active-training
    block — see ml_jobs.BLOCK_SECONDS), so it adds in directly. Training
    therefore draws from the SAME monthly credit pool as labelling /
    uploads / storage, which is why it's summed here rather than gated
    separately."""
    labelled = labelled_this_month / LABELLED_IMAGES_PER_CREDIT
    uploaded = uploaded_this_month / UPLOADED_IMAGES_PER_CREDIT
    stored = math.ceil(stored_now / STORED_IMAGES_PER_CREDIT_PER_MONTH)
    return labelled + uploaded + stored + max(0.0, float(training_credits_this_period))


# ─── Plan limits ──────────────────────────────────────────────────────────────

class PlanLimits(TypedDict):
    creditsPerMonth: int
    imagesLabelledPerMonth: int
    projects: int
    allowsPrivateProjects: bool


def _build_limits(*, credits: int, projects: int, private_projects: bool) -> PlanLimits:
    return {
        "creditsPerMonth": credits,
        # Legacy flat cap derived from credits so older call sites
        # (FE display) keep getting a number, in lockstep with the FE.
        "imagesLabelledPerMonth": credits * LABELLED_IMAGES_PER_CREDIT,
        "projects": projects,
        "allowsPrivateProjects": private_projects,
    }


PLAN_LIMITS: dict[PlanId, PlanLimits] = {
    "free":       _build_limits(credits=5,      projects=5,      private_projects=False),
    "beta":       _build_limits(credits=50,     projects=200,    private_projects=True),
    "pro_50":     _build_limits(credits=50,     projects=200,    private_projects=True),
    "pro_150":    _build_limits(credits=150,    projects=200,    private_projects=True),
    "pro_500":    _build_limits(credits=500,    projects=200,    private_projects=True),
    "mega":       _build_limits(credits=10_000, projects=10_000, private_projects=True),
    "enterprise": _build_limits(credits=10_000, projects=10_000, private_projects=True),
}


def limits_for(plan_id: str | None) -> PlanLimits:
    return PLAN_LIMITS[plan_for(plan_id)]


# Founder / internal-account credit overrides (username.lower() → credits/mo).
# Mirror of the FE's limitsForUser. Keeping the override on the SERVER too
# means a future deploy of the FE can't lose it without the backend
# immediately re-capping the account.
_CREDIT_OVERRIDES: dict[str, int] = {
    "hamish": 10_000,
    "mukund": 1_000_000,
}


def limits_for_user(plan_id: str | None, username: str | None) -> PlanLimits:
    base = limits_for(plan_id)
    credits = _CREDIT_OVERRIDES.get((username or "").strip().lower())
    if credits is not None:
        return {
            **base,
            "creditsPerMonth": credits,
            "imagesLabelledPerMonth": credits * LABELLED_IMAGES_PER_CREDIT,
        }
    return base


# ─── Effective plan resolution ────────────────────────────────────────────────

def _parse_beta_expires(beta_expires_at: str | None) -> float | None:
    """ISO-8601 string (with or without ms / Z) → unix-seconds. Returns
    None if absent / unparseable so the caller can treat that as "no
    beta override"."""
    if not beta_expires_at:
        return None
    s = beta_expires_at.strip()
    if not s:
        return None
    # datetime.fromisoformat accepts the "...Z" suffix from Py 3.11+;
    # for older interpreters strip it explicitly to keep parity.
    if s.endswith("Z"):
        s = s[:-1] + "+00:00"
    try:
        return datetime.fromisoformat(s).timestamp()
    except Exception:
        return None


def effective_plan_for(
    plan_id: str | None,
    status: str | None,
    beta_expires_at: str | None = None,
) -> PlanId:
    """What the user is *effectively* on right now, taking subscription
    health into account. Mirror of the FE's effectivePlanFor.

      - Mega / Enterprise beat every other tier (admin-granted, no clock)
      - Active beta window overrides the stored tier
      - A stored Pro tier with a non-paid Stripe status drops to Free
      - A stored Beta with a lapsed/missing window drops to Free
    """
    stored = plan_for(plan_id)
    if stored == "mega":
        return "mega"
    if stored == "enterprise":
        return "enterprise"
    beta_exp = _parse_beta_expires(beta_expires_at)
    if beta_exp is not None and beta_exp > time.time():
        return "beta"
    if stored == "free":
        return "free"
    if stored == "beta":
        return "free"  # lapsed beta without active timestamp
    return stored if (status or "").lower() in PAID_STATUSES else "free"


# ─── Cycle window ─────────────────────────────────────────────────────────────

def _shift_months(d: datetime, delta: int) -> datetime:
    """Add `delta` months to `d`, clamping the day-of-month to the
    target month's length (e.g. Jan 31 + 1 → Feb 28/29)."""
    target_year = d.year + (d.month - 1 + delta) // 12
    target_month = (d.month - 1 + delta) % 12 + 1
    # Last day of target month.
    if target_month == 12:
        last = 31
    else:
        next_first = datetime(target_year, target_month + 1, 1, tzinfo=d.tzinfo)
        last = (next_first.replace(day=1) - _one_day()).day
    day = min(d.day, last)
    return d.replace(year=target_year, month=target_month, day=day)


def _one_day():
    from datetime import timedelta
    return timedelta(days=1)


def _next_anchor_date(anchor: datetime, now: datetime) -> datetime:
    d = anchor
    while d <= now:
        d = _shift_months(d, 1)
    return d


def cycle_window(
    *,
    effective_plan: PlanId,
    subscription_current_period_end: datetime | None,
    signup_date: datetime,
    now: datetime | None = None,
) -> tuple[datetime, datetime]:
    """Resolve the current billing cycle's [start, end). Pro users
    anchor on subscriptionCurrentPeriodEnd (Stripe's authoritative
    next-charge time); everyone else anchors on their signup day-of-
    month. Same rules as the FE's cycleWindow."""
    now = now or datetime.now(timezone.utc)
    if is_pro_plan(effective_plan) and subscription_current_period_end is not None:
        end = subscription_current_period_end
    else:
        end = _next_anchor_date(signup_date, now)
    start = _shift_months(end, -1)
    return start, end


# ─── Convenience aggregation ──────────────────────────────────────────────────

class CreditState(TypedDict):
    plan: PlanId
    limits: PlanLimits
    credits_used: float
    credits_remaining: float
    over_credits: bool


def evaluate_credits(
    *,
    username: str | None,
    plan_id: str | None,
    status: str | None,
    beta_expires_at: str | None,
    labelled_this_month: int,
    uploaded_this_month: int,
    stored_now: int,
    training_blocks_this_period: int = 0,
) -> CreditState:
    """One-shot resolution: take the JWT claims + usage counters,
    return the effective plan + computed credit ledger + over-cap flag.
    The HTTP gate (`enforce_credits_or_402`) reads this to decide
    whether to allow an AI spend.

    `training_blocks_this_period` is the count of 15-min active-training
    blocks already billed this cycle (1 block = 1 credit)."""
    eff = effective_plan_for(plan_id, status, beta_expires_at)
    limits = limits_for_user(eff, username)
    used = credits_used(
        labelled_this_month=labelled_this_month,
        uploaded_this_month=uploaded_this_month,
        stored_now=stored_now,
        training_credits_this_period=float(training_blocks_this_period),
    )
    allowed = float(limits["creditsPerMonth"])
    return {
        "plan": eff,
        "limits": limits,
        "credits_used": used,
        "credits_remaining": max(0.0, allowed - used),
        "over_credits": allowed > 0 and used >= allowed,
    }


__all__: Iterable[str] = (
    "PlanId",
    "PRO_TIER_IDS",
    "PAID_STATUSES",
    "PlanLimits",
    "PLAN_LIMITS",
    "LABELLED_IMAGES_PER_CREDIT",
    "UPLOADED_IMAGES_PER_CREDIT",
    "STORED_IMAGES_PER_CREDIT_PER_MONTH",
    "plan_for",
    "is_pro_plan",
    "credits_used",
    "limits_for",
    "limits_for_user",
    "effective_plan_for",
    "cycle_window",
    "evaluate_credits",
    "CreditState",
)

File: ui/lib/plans.ts
// Single source of truth for plan limits + credit allocations. Both
// the profile page and the create/label gates read from here so the
// displayed quotas and the enforced quotas can't drift apart.

// Plan IDs. Three Pro tiers replace the single old Pro plan so the
// user can pick a price / credit allocation that matches their
// volume. `beta` and `mega` stay for backward compatibility with
// existing accounts; new signups can't reach them via Stripe
// checkout (beta = code redemption; mega = admin-granted).
//
// `enterprise` is the contact-sales destination, no Stripe product
// is created for it.
export type PlanId =
  | "free"
  | "beta"
  | "pro_50"
  | "pro_150"
  | "pro_500"
  | "mega"
  | "enterprise";

// Pro tier IDs. Used by the FE pricing-card selector + checkout
// payload validation. Order matters: array index drives the default
// "selected" tier on the pricing card (mid-tier).
export type ProTierId = "pro_50" | "pro_150" | "pro_500";
export const PRO_TIER_IDS: ProTierId[] = ["pro_50", "pro_150", "pro_500"];

export type PlanLimits = {
  // Credits per calendar month. One credit covers ~100 labelled
  // images, ~800 uploaded images, or ~400 stored images for the
  // month. See CREDIT_COSTS below.
  creditsPerMonth: number;
  // Hard cap on auto-labelled images per calendar month. Derived
  // from creditsPerMonth × CREDIT_COSTS.labelledImagesPerCredit so
  // existing call-sites that gate on a flat image cap keep working
  // during the credit-system rollout. Once everything reads from
  // `creditsPerMonth` directly we can drop this.
  imagesLabelledPerMonth: number;
  // Hard cap on projects the user can own (public + private combined).
  projects: number;
  // Whether the plan permits private projects. Free is public-only.
  allowsPrivateProjects: boolean;
};

export type PlanMeta = {
  name: string;
  // Display label, "Pro" for all three pro tiers, distinct names
  // for free/beta/mega/enterprise.
  shortName: string;
  // Monthly GBP. null for plans that don't have a fixed price
  // (mega = admin grant, enterprise = custom, beta = free).
  priceGbp: number | null;
  limits: PlanLimits;
};

// Build a PlanLimits from just credits + projects + private flag.
// Derives the legacy imagesLabelledPerMonth field so older call
// sites still resolve a number without us repeating the arithmetic
// at every entry. Kept private to plans.ts.
function makeLimits(opts: {
  credits: number;
  projects: number;
  privateProjects: boolean;
}): PlanLimits {
  return {
    creditsPerMonth: opts.credits,
    imagesLabelledPerMonth: opts.credits * 100, // 1 credit = 100 labels
    projects: opts.projects,
    allowsPrivateProjects: opts.privateProjects,
  };
}

export const PLANS: Record<PlanId, PlanMeta> = {
  free: {
    name: "Free",
    shortName: "Free",
    priceGbp: 0,
    limits: makeLimits({ credits: 5, projects: 5, privateProjects: false }),
  },
  // Free-of-charge time-boxed tier handed out via beta codes. Matches
  // the entry Pro tier so beta testers can stress the product;
  // auto-downgrades to Free 30 days after redemption.
  beta: {
    name: "Beta",
    shortName: "Beta",
    priceGbp: 0,
    limits: makeLimits({ credits: 50, projects: 200, privateProjects: true }),
  },
  pro_50: {
    name: "Pro",
    shortName: "Pro · 50",
    priceGbp: 29,
    limits: makeLimits({ credits: 50, projects: 200, privateProjects: true }),
  },
  pro_150: {
    name: "Pro",
    shortName: "Pro · 150",
    priceGbp: 69,
    limits: makeLimits({ credits: 150, projects: 200, privateProjects: true }),
  },
  pro_500: {
    name: "Pro",
    shortName: "Pro · 500",
    priceGbp: 179,
    limits: makeLimits({ credits: 500, projects: 200, privateProjects: true }),
  },
  // Admin-granted only (no Stripe checkout). For internal users +
  // hand-picked partners. Effectively-unlimited limits.
  mega: {
    name: "Mega",
    shortName: "Mega",
    priceGbp: null,
    limits: makeLimits({ credits: 10_000, projects: 10_000, privateProjects: true }),
  },
  // Contact-sales tier. Limits are placeholders, actual quota is
  // negotiated per customer and applied via mega plus a separate
  // ledger entry.
  enterprise: {
    name: "Enterprise",
    shortName: "Enterprise",
    priceGbp: null,
    limits: makeLimits({ credits: 10_000, projects: 10_000, privateProjects: true }),
  },
};

// Credit cost table. The pricing page's usage table renders straight
// from this object so the doc and the gate can't disagree.
export const CREDIT_COSTS = {
  labelledImagesPerCredit: 100,
  uploadedImagesPerCredit: 800,
  storedImagesPerCreditPerMonth: 400,
} as const;

// Compute credits used in a billing window from raw counters. Used
// by both the profile page and the backend's usage endpoint so the
// display + the gate share the same arithmetic.
//
// Storage is rounded UP because a partial-month occupation still
// burns a credit-month, the user occupies that slot.
export function creditsUsed(opts: {
  labelledThisMonth: number;
  uploadedThisMonth: number;
  storedNow: number;
}): number {
  const labelled = opts.labelledThisMonth / CREDIT_COSTS.labelledImagesPerCredit;
  const uploaded = opts.uploadedThisMonth / CREDIT_COSTS.uploadedImagesPerCredit;
  const stored = Math.ceil(opts.storedNow / CREDIT_COSTS.storedImagesPerCreditPerMonth);
  return labelled + uploaded + stored;
}

// Resolve a stored plan-id string into the canonical PlanId enum.
// Tolerates the legacy "pro" string by mapping it to the entry tier
// so accounts created before the three-tier split don't lose their
// subscription.
export function planFor(planId: string | null | undefined): PlanId {
  if (planId === "mega") return "mega";
  if (planId === "enterprise") return "enterprise";
  if (planId === "pro_50") return "pro_50";
  if (planId === "pro_150") return "pro_150";
  if (planId === "pro_500") return "pro_500";
  if (planId === "pro") return "pro_50"; // legacy pre-split → entry tier
  if (planId === "beta") return "beta";
  return "free";
}

export function isProPlan(id: PlanId): id is ProTierId {
  return id === "pro_50" || id === "pro_150" || id === "pro_500";
}

export function limitsFor(planId: string | null | undefined): PlanLimits {
  return PLANS[planFor(planId)].limits;
}

// Per-user limit override. Resolves the plan's normal limits, then
// applies any account-level grants on top. Used by the /api/users/usage
// route so the FE displays the same allowance the backend honours.
//
// Founder / internal-account credit overrides (username.lower() → credits/mo).
// Mirror of the backend's _CREDIT_OVERRIDES in plans.py — keeping it on both
// sides means a deploy of either can't silently lose the override.
//
// Account overrides currently:
//   hamish - founder account (10,000 credits/month)
//   mukund - founder account (1,000,000 credits/month)
const CREDIT_OVERRIDES: Record<string, number> = {
  hamish: 10_000,
  mukund: 1_000_000,
};

export function limitsForUser(
  planId: string | null | undefined,
  username: string | null | undefined,
): PlanLimits {
  const base = limitsFor(planId);
  const credits = CREDIT_OVERRIDES[(username ?? "").trim().toLowerCase()];
  if (credits != null) {
    return {
      ...base,
      creditsPerMonth: credits,
      imagesLabelledPerMonth: credits * CREDIT_COSTS.labelledImagesPerCredit,
    };
  }
  return base;
}

export function planName(planId: string | null | undefined): string {
  return PLANS[planFor(planId)].name;
}

// Stripe statuses that count as a paid subscription. Everything else
// (past_due, unpaid, canceled, incomplete, incomplete_expired, paused)
// drops the user back to the free quota until they fix their payment.
const PAID_STATUSES = new Set(["active", "trialing"]);

// Compute the start + end of the current usage cycle for a user.
//
//   - Pro user with a known `subscriptionCurrentPeriodEnd`: the cycle
//     ends on that date (the day Stripe will charge them next), and
//     starts one month earlier. This way a Pro user who paid on the
//     15th sees their quota reset on the 15th, not on the 1st.
//   - Everyone else (Free, or a Pro user we don't yet have a period
//     end for): anchor on the day-of-month they signed up. The cycle
//     ends on the next signup-day-anniversary in the future and
//     starts one month before that.
//
// Returns ISO timestamps (truncated to the second) suitable for
// passing straight to the FastAPI usage endpoint.
export function cycleWindow(opts: {
  effectivePlan: PlanId;
  subscriptionCurrentPeriodEnd: Date | null;
  signupDate: Date;
  now?: Date;
}): { start: string; end: string } {
  const now = opts.now ?? new Date();
  let end: Date;
  if (isProPlan(opts.effectivePlan) && opts.subscriptionCurrentPeriodEnd) {
    end = new Date(opts.subscriptionCurrentPeriodEnd);
  } else {
    // Free, Beta, Mega and Enterprise all anchor on the signup
    // day-of-month, Beta doesn't need a Stripe-style sliding
    // window because its lifecycle is driven by `betaExpiresAt`
    // instead.
    end = nextAnchorDate(opts.signupDate, now);
  }
  const start = shiftMonths(end, -1);
  return {
    start: start.toISOString().replace(/\.\d{3}Z$/, "Z"),
    end: end.toISOString().replace(/\.\d{3}Z$/, "Z"),
  };
}

// Roll a date forward in 1-month steps until it's strictly after `now`.
// Handles month-length differences by clamping to the last day of the
// month (e.g. signed up Jan 31 → cycles end Feb 28/29, then Mar 31).
function nextAnchorDate(anchor: Date, now: Date): Date {
  let d = new Date(anchor.getTime());
  while (d <= now) {
    d = shiftMonths(d, 1);
  }
  return d;
}

function shiftMonths(d: Date, delta: number): Date {
  const day = d.getUTCDate();
  const target = new Date(Date.UTC(
    d.getUTCFullYear(),
    d.getUTCMonth() + delta,
    1,
    d.getUTCHours(),
    d.getUTCMinutes(),
    d.getUTCSeconds(),
    0,
  ));
  const lastDayOfTarget = new Date(Date.UTC(
    target.getUTCFullYear(),
    target.getUTCMonth() + 1,
    0,
  )).getUTCDate();
  target.setUTCDate(Math.min(day, lastDayOfTarget));
  return target;
}

// What the user is *effectively* on right now, taking subscription
// health into account. `subscriptionPlan` records what they paid for
// most recently; this function is what the rest of the app should
// read when deciding limits. A failed renewal therefore re-engages
// the free-tier quota gates without us touching the DB column.
//
// Beta is time-boxed via `betaExpiresAt` and sits *over* whatever
// plan the user already has (except Mega/Enterprise): a Pro user who
// redeems a beta code reads as Beta for 30 days, then drops back to
// Pro when the window lapses.
export function effectivePlanFor(
  planId: string | null | undefined,
  status: string | null | undefined,
  betaExpiresAt?: string | Date | null,
): PlanId {
  const stored = planFor(planId);

  // Mega + Enterprise beat every other tier, admin-granted, no clock.
  if (stored === "mega") return "mega";
  if (stored === "enterprise") return "enterprise";

  // Active beta window overrides the stored tier. A user's stored
  // plan during the window depends on the code they redeemed: the
  // backer/Pro codes set subscriptionPlan="pro_50" and read AS Pro,
  // the legacy beta code keeps subscriptionPlan="beta" and reads AS
  // Beta. After expiry the stored plan takes over again (Free for
  // code-only users; the real Stripe tier for paying users).
  if (betaExpiresAt) {
    const expiry = betaExpiresAt instanceof Date ? betaExpiresAt : new Date(betaExpiresAt);
    if (Number.isFinite(expiry.getTime()) && expiry.getTime() > Date.now()) {
      // Stored Pro tier (from a Pro-grade redeem code OR a Stripe
      // subscription) wins over "beta" because Pro is the strictly
      // higher tier — beta has identical limits to pro_50 but a
      // lower-rank label, and we don't want to downgrade a Pro
      // user's badge during an active beta window.
      if (isProPlan(stored)) return stored;
      return "beta";
    }
  }

  if (stored === "free") return "free";
  // A stored "beta" with a lapsed (or missing) timestamp drops to Free.
  if (stored === "beta") return "free";
  // Pro tiers need a healthy Stripe subscription.
  return status && PAID_STATUSES.has(status) ? stored : "free";
}
